### PR TITLE
Add index_prefix option to text fields

### PR DESCRIPTION
**Pull request by romseygeek**
This adds the ability to index term prefixes into a hidden subfield, enabling prefix queries to be run without multitermquery rewrites.  The subfield reuses the analysis chain of its parent text field, appending an EdgeNGramTokenFilter.  It can be configured with minimum and maximum ngram lengths.  Query terms with lengths outside this min-max range fall back to using prefix queries against the parent text field.

The mapping looks like this:
```
"my_text_field" : {
    "type" : "text",
    "analyzer" : "english",
    "index_prefix" : { "min_chars" : 1, "max_chars" : 10 }
}
```

### CODE REVIEW

**Comment by romseygeek** [comment]
This is still a work-in-progress, and needs more comprehensive tests + docs, but I'd like to get some feedback on whether or not this is a sensible implementation.

File: server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java [hunk 1]
```diff
@@ -273,11 +329,27 @@ public void setFielddataMinSegmentSize(int fielddataMinSegmentSize) {
...
         public String typeName() {
             return CONTENT_TYPE;
         }
 
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (prefixAnalyzer == null || prefixAnalyzer.accept(value.length()) == false) {
+                return super.prefixQuery(value, method, context);
+            }
+            TermQuery q = new TermQuery(new Term(name() + "._prefix", indexedValueForSearch(value)));
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java]: I don't think anything prevents a user from creating an explicit field with the same name?

Comment (romseygeek) [reply server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java]: Not yet, no.  Do we have a way of reserving field names elsewhere?

**Comment by jpountz** [comment]
I think you are on the right track.

@rjernst raises a good point that there could be conflicts if a user configures a multi-field that also has `_prefix` as a name.

> Do we have a way of reserving field names elsewhere?

I don't think we do. We only reserve fields that start with `_` on the top level. I think the only restriction that we put on inner levels is that fields cannot contain a dot. Thinking out loud: would calling the field `${field_name}..prefix` be a viable option? Such a field name should be illegal for regular fields.

**Comment by romseygeek** [comment]
Closing in favour of #28290

**Pull request closed without merging**